tutorial-review.md

```diff
@@ -0,0 +1,1186 @@
+├── routes
+│   ├── index.js
+│   └── users.js
+└── views
+    ├── error.pug
+    ├── index.pug
+    └── layout.pug
+└── client
+    ├── package.json
+    ├── yarn.lock
+    ├── public
+    └── src
+        ├── App.js
+        ├── App.css
+        ├── App.test.js
+        ├── index.js
+        ├── index.css
+        ├── logo.svg
+        ├── reportWebVitals.js
+        └── setupTests.js
+
+```
+
+### Create a database on Mongo Atlas
+
+The easiest way to setup our MongoDB database is to rely on a cloud service such as [Mongo Atlas](https://www.mongodb.com/cloud/atlas/lp/try2). They host databases on AWS, Google Cloud, Azure and provide several options to operate and scale MongoDB painlessly.
+
+
+In the plan page, choose the "shared" plan which starts for free:
+[mongo-plan]
+
+Next, choose the cloud provider and region, leave the default values for "cluster tier", "additional settings" and choose a cluster name, "mongo-blog-db".
+
+
+
+
+In the dashboard "quickstart" go ahead and create a username and password as we will need that to connect the database later on. Plus, we can restrict IP access to the database. This is very important measure when going to production with a real app but for the purpose of the tutorial we can simply add `0.0.0.0/0` to allow access from any IP.
+
+
+The database will successfully deploy in a few minutes, in the meantime we can click on the "connect" button, choose "connect your application" and copy the connection string.
+
+[mongo-connection]
+A typical connection string should look  like this:
+
+```
+mongodb+srv://samuele:<password>@mongo-client-db.r5bv5.mongodb.net/myFirstDatabase?retryWrites=true&w=majority
+```
+
+Move back the codebase, open `app.js` to require mongoose, connect it to the database by using the connection string and recover from potential errors:
+
+```javascript
+const mongoose = require('mongoose');
+const CONNECTION_STRING = `process.env.CONNECTION_STRING` // env variable we can defined in the package.json containing the connection string and pwd
+
+// setup connection to mongo
+mongoose.connect(CONNECTION_STRING);
+const db = mongoose.connection;
+// recover from errors
+db.on('error', console.error.bind(console, 'connection error:'));
+```
+
+Since we decided to get the connection string as an environment variable, to test it in development we can add it to the package.json located in the root directory:
+
+```bash
+{
+  "name": "mongo-blog",
+  "version": "0.0.0",
+  "private": true,
+  "scripts": {
+    "dev": "nodemon ./bin/www",
+    "start": "node ./bin/www",
+    "build-client": "cd ./client && yarn build"
+  },
+  "dependencies": {
+    "cookie-parser": "~1.4.4",
+    "cors": "^2.8.5",
+    "debug": "~2.6.9",
+    "express": "~4.16.1",
+    "http-errors": "~1.6.3",
+    "jade": "~1.11.0",
+    "mongoose": "^6.2.3",
+    "morgan": "~1.9.1"
+  },
+  "devDependencies": {
+    "nodemon": "^2.0.15"
+  },
++ "nodemonConfig": {
++   "env": {
++     "MONGO_PWD": "YOUR_MONGO_USER_PWD"
```

Comment:
Is it ok to hardcode the password here?

Reply:
For the purpose of the tutorial I could say yes, definitely not for an app going to production

Reply:
@alisdairbr just removed this piece of code